File: csharp/Src/Shapes/Base/Commons/BaseShape.cs
namespace Shapes.Base.Commons;

public abstract class BaseShape {
    private readonly List<Dictionary<string, double>> _dimensions;

    /// <summary>
    /// Initialize the BaseShape with a list of dictionaries representing dimensions.
    /// Each dictionary should have a single key-value pair, e.g.,
    /// <c>[{"width": 12}, {"length": 15}]</c>.
    /// </summary>
    protected BaseShape(List<Dictionary<string, double>> dimensions) {
        _dimensions = dimensions;
    }

    public List<Dictionary<string, double>> Dimensions {
        get => _dimensions;
    }

    public int ShowAvailableDimensionCount() {
        return _dimensions.Count;
    }

    public static bool ValidateShape() {
        return true;
    }

    public abstract void PrintDescription();
}

File: scala/src/main/scala/shapes/base/commons/BaseShape.scala
package shapes.base.commons

abstract class BaseShape(private var _dimensions: List[Map[String, Double]] = List()) {
    def dimensions: List[Map[String, Double]] = {
        _dimensions
    }

    def showAvailableDimensionCount: Int = {
        _dimensions.size
    }

    def printDescription(): Unit
}

object BaseShape {
    def validateShape: Boolean = {
        true
    }
}
